File: Arrays/Day 4/Java/solution.java
class Solution {
    
    static void reverse(int[] arr,int start, int end) {
        while(start < end) {
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }    
    }
    
    static void rotateArr(int arr[], int d) {
        int len = arr.length;
        if(d == 0 || d==len) {
            return;
        }
        
        if(d > len) {
            d = d%len;
        }
        
        //Reverse the first d elements
        reverse(arr,0,d-1);
        
        //Reverse the remaining elements
        reverse(arr,d,len-1);
        
        //Reverse the entire array
        reverse(arr,0,len-1);
    }
}

File: Arrays/Day 4/Python/solution.py
class Solution(object):
    def algo(self,nums,start,end):
        while(start < end):
            temp = nums[start]
            nums[start] = nums[end]
            nums[end] = temp
            start+=1
            end-=1

    def rotate(self, nums, k):
        if(k == 0 or k==len(nums)):
            return
        
        if(k>len(nums)):
            k=k%len(nums)

        length = len(nums)

        self.algo(nums,0,length-1)

        self.algo(nums,0,k-1)

        self.algo(nums,k,length-1)
        
